org.eclipse.buildship.core/src/main/java/org/eclipse/buildship/core/workspace/internal/GradleBuilderAndNatureMergingStrategy.java

```diff
@@ -0,0 +1,87 @@
+/*
+ * Copyright (c) 2017 the original author or authors.
+ * All rights reserved. This program and the accompanying materials
+ * are made available under the terms of the Eclipse Public License v1.0
+ * which accompanies this distribution, and is available at
+ * http://www.eclipse.org/legal/epl-v10.html
+ */
+
+package org.eclipse.buildship.core.workspace.internal;
+
+import java.util.Set;
+
+import com.google.common.collect.Sets;
+
+/**
+ * Algorithm for updating the Gradle natures and builders that considers the existing user setup.
+ * <p/>
+ * The algorithm will update existing elements with the new model such that only the managed
+ * elements will be deleted. As of the ordering, the model elements are always first, followed by
+ * the user-defined entries.
+ *
+ * @author Donat Csikos
+ */
+final class GradleBuilderAndNatureMergingStrategy {
+
+    /**
+     * Calculates the updated state.
+     *
+     * @param current elements currently defined on the project
+     * @param model elements defined in the Gradle model
+     * @param managed elements managed by Buildship
+     * @return the description of the updated state
+     */
+    public static <T> Result<T> calculate(Set<T> current, Set<T> model, Set<T> managed) {
```

Comment:
The `Sets` class in Guava has union/intersection/difference methods that might be useful here.